datafusion/core/src/datasource/memory.rs

```diff
@@ -73,6 +75,26 @@ impl MemTable {
         }
     }
 
+    /// Create a new in-memory table from the record batches.
+    /// In case of empty table, the schema is inferred from the input plan.
+    pub fn new_not_registered(
```

Comment:
I think this code effectively ignores the `schema` argument if any of `partitions` has a RecordBatch and uses that RecordBatch's schema instead. 

I think this is a surprising behavior and might mask errors in the future